web/modules/custom/collation_fixer/src/CollationFixer.php

```diff
@@ -0,0 +1,199 @@
+
+      if (($schema_charset != $db_charset) || ($schema_collation != $db_collation)) {
+        $wrong_collations[] = new TableCollation($table_name, $db_collation, $db_charset, $schema_collation, $schema_charset);
+      }
+    }
+
+    return $wrong_collations;
+  }
+
+  /**
+   * Fix tables collations.
+   *
+   * @param string $table
+   *   Name of a table to fix collation on. Defaults to all tables.
+   *
+   * @return bool
+   *   TRUE if collations where changed successfully.
+   */
+  public function fixCollation(?string $table = NULL) {
+    $wrongCollations = $this->checkCollation($table);
+
+    $status = TRUE;
+
+    foreach ($wrongCollations as $tableCollation) {
+
+      // Alter character set and collation of table definition.
+      $query = $this->connection->query(
+        // Placeholders are intentionally not used there. It does not seem as
+        // if is it supported.
+        "ALTER TABLE {$tableCollation->table} CHARACTER SET {$tableCollation->expectedCharset} COLLATE {$tableCollation->expectedCollation}"
+      );
+      if ($query instanceof StatementInterface) {
+        $result = $query->execute();
+        $status = $status && $result;
+      }
+
+      // Alter character set and collation of table data.
+      $query = $this->connection->query(
+        // Placeholders are intentionally not used there. It does not seem as
+        // if is it supported.
+        "ALTER TABLE {$tableCollation->table} CONVERT TO CHARACTER SET {$tableCollation->expectedCharset} COLLATE {$tableCollation->expectedCollation}"
+      );
+      if ($query instanceof StatementInterface) {
+        $result = $query->execute();
+        $status = $status && $result;
+      }
+    }
+
+    return $status;
+  }
+
+  /**
+   * Returns all available schemas ie. table names.
+   *
+   * @param ?string $table
+   *   A specific table name to check.
+   *
+   * @return string[]
+   *   An array of table names.
+   */
+  private function getSchema(?string $table = NULL): array {
+    $schemas = array_map(function (string $module) {
+      if ($this->moduleHandler->loadInclude($module, 'install')) {
+        return $this->moduleHandler->invoke($module, 'schema') ?? [];
+      }
+      else {
+        return [];
+      }
+    }, array_keys($this->moduleHandler->getModuleList()));
+    $moduleSchemas = array_merge(...$schemas);
+
+    $entitySchemas = [];
+    foreach ($this->entityTypeManager->getDefinitions() as $entityType) {
+      // Only list content entity types using SQL storage.
+      $entityStorage = $this->entityTypeManager->getStorage($entityType->id());
+      if ($entityStorage instanceof SqlEntityStorageInterface) {
+        foreach ($this->fieldManager->getFieldStorageDefinitions($entityType->id()) as $field) {
+          $entitySchemas = array_merge($entitySchemas, $entityStorage->getTableMapping()->getAllFieldTableNames($field->getName()));
+        }
+      }
+    }
```

Comment:
This approach does not seem to get *all* tables used by the installation. I am not use if this is a problem and if so what to do about it.

If we query the database for all tables we risk getting tables that are not part of the current installation.

Reply:
Hmm… only tables for entities? Maybe just document it as a feature/limitation? It will mostly be useful content, and content usually resides in entities.